File: Jack.Data/Repository/GenericDao.vb
Imports NHibernate
Imports System.Collections.Generic

''' <summary>
''' Classe genérica para acesso a dados utilizando NHibernate. Como utiliza generics, é possível persistir
''' qualquer objeto, bastando definir os tipos a serem usados no lugar de Tipo e ID.
''' <example>
''' Dim usuario As New Usuario
''' Dim usuarioDAO As New GenericDAO(Of Usuario, Integer) 
''' 
''' usuario.Nome = "Rogério"
''' usuario.Telefone = "19 32612649"
''' usuario.Email = "bragil@gmail.com"
''' 
''' If usuarioDAO.Insert(usuario) Then
'''      Console.WriteLine("Sucesso!")
''' Else
'''      Console.WriteLine("Erro ao inserir registro!")
''' End If
''' </example>
''' </summary>
''' <typeparam name="Tipo">Tipo do objeto persistível</typeparam>
''' <typeparam name="ID">Tipo de dado do identificador único (chave primária)</typeparam>
''' 
Public Class GenericDao(Of Tipo, ID)
    Implements IGenericDao(Of Tipo, ID), IDisposable

    Private oSession As ISession
    Private oTransaction As ITransaction
    Private blnHasTransaction As Boolean
    'Private oNhTemplate As 

    ''' <summary>
    ''' Construtor, obtém a sessão do NHibernateHelper e inicia a transação.
    ''' </summary>
    Public Sub New(Optional ByVal blnWithTransaction As Boolean = False)
        Try
            oSession = NHibernateHelper.GetCurrentSession(GetType(Tipo).Assembly)
            blnHasTransaction = blnWithTransaction
            If blnWithTransaction Then
                oTransaction = oSession.BeginTransaction()
            End If
        Catch ex As Exception
            Throw ex
        End Try
    End Sub

    ''' <summary>
    ''' Persiste um novo objeto.
    ''' </summary>
    ''' <param name="Obj">Objeto a ser persistido</param>
    ''' <returns>True para sucesso, False para erro.</returns>
    Public Overridable Function Insert(ByVal Obj As Tipo) As Boolean Implements IGenericDao(Of Tipo, ID).Insert
        Dim oRetorno As Object
        Try
            oRetorno = oSession.Save(Obj)
            Finish()
        Catch ex As Exception
            Return False
        End Try
        Return True
    End Function

    ''' <summary>
    ''' Atualiza um objeto existente.
    ''' </summary>
    ''' <param name="Obj">Objeto a ser atualizado</param>
    ''' <returns>True para sucesso, False para erro.</returns>
    Public Overridable Function Update(ByVal Obj As Tipo) As Boolean Implements IGenericDao(Of Tipo, ID).Update
        Try
            oSession.Update(Obj)
            Finish()
        Catch ex As Exception
            Return False
        End Try
        Return True
    End Function

    ''' <summary>
    ''' Exclui um objeto existente.
    ''' </summary>
    ''' <param name="Obj">Objeto a ser atualizado</param>
    ''' <returns>True para sucesso, False para erro.</returns>
    Public Overridable Function Delete(ByVal Obj As Tipo) As Boolean Implements IGenericDao(Of Tipo, ID).Delete
        Try
            oSession.Delete(Obj)
            Finish()
        Catch ex As Exception
            Return False
        End Try
        Return True
    End Function

    ''' <summary>
    ''' Recupera um objeto persistido na base de dados.
    ''' </summary>
    ''' <param name="Identifier">Chave primária do objeto</param>
    ''' <returns>Objeto</returns>
    Public Overridable Function Find(ByVal Identifier As ID) As Tipo Implements IGenericDao(Of Tipo, ID).Find
        Dim oTipo As Tipo
        Try

            oTipo = CType(oSession.Load(GetType(Tipo), Identifier), Tipo)
            Return oTipo
        Catch ex As Exception
            Throw ex
        End Try
    End Function

    ''' <summary>
    ''' Retorna todos os objetos persistidos na base de dados.
    ''' </summary>
    ''' <returns>Coleção de objetos</returns>
    Public Overridable Function LoadAll() As IList(Of Tipo) Implements IGenericDao(Of Tipo, ID).LoadAll
        Dim lTipo As IList(Of Tipo)
        Try
            'Prepare()
            lTipo = oSession.CreateCriteria(GetType(Tipo)).List(Of Tipo)()
            Return lTipo
        Catch ex As Exception
            Throw ex
        End Try
    End Function

    ''' <summary>
    ''' Confirma a transação corrente (commit)
    ''' </summary>
    Protected Overridable Sub Commit()
        oTransaction.Commit()
    End Sub

    ''' <summary>
    ''' Desfaz a transação corrente (rollback)
    ''' </summary>
    Protected Overridable Sub BeginTransaction()
        oTransaction = oSession.BeginTransaction()
    End Sub

    ''' <summary>
    ''' Begin Trans
    ''' </summary>
    ''' <remarks></remarks>
    Protected Overridable Sub Rollback()
        oTransaction.Rollback()
    End Sub

    ''' <summary>
    ''' Reinicia a Sessão
    ''' </summary>
    ''' <remarks></remarks>
    Protected Overridable Sub Finish()
        oSession.Flush()
        oSession.Clear()
    End Sub

    ''' <summary>
    ''' Verifica se há ou não transação
    ''' </summary>
    ''' <value>Boolean</value>
    ''' <returns>Boolean</returns>
    ''' <remarks></remarks>
    Public Property HasTransaction() As Boolean
        Get
            Return blnHasTransaction
        End Get
        Set(ByVal value As Boolean)
            blnHasTransaction = value
        End Set
    End Property

    ''' <summary>
    ''' Retorna dados através de uma query
    ''' </summary>
    ''' <param name="strQuery">Query</param>
    ''' <param name="strValue">Valor a Filtrar</param>
    ''' <param name="strParName">Nome do Parametro</param>
    ''' <returns>Lista Tipada</returns>
    ''' <remarks></remarks>
    Function LoadByQuery(ByVal strQuery As String, _
                         ByVal strParName As String, _
                         ByVal strValue As String) As IList(Of Tipo)

        Dim oReturn As IList(Of Tipo)
        Dim oQuery As NHibernate.IQuery

        Try
            oQuery = oSession.CreateQuery(strQuery) _
            .SetString(strParName, strValue)
            oReturn = oQuery.List(Of Tipo)()

        Catch ex As Exception
            oReturn = Nothing
            Throw ex
        End Try

        Return oReturn

    End Function


#Region "other Methods"

    'public virtual List<T> LoadByQuery(string hsqlQuery, object[] values)
    '{

    '    List<T> listReturn = new List<T>();
    '    IList oArray = null;

    '    oArray = nhTemplate.Find(hsqlQuery, values) as IList;
    '    listReturn = new List<T>();
    '    foreach (T oData in oArray)
    '    {
    '        listReturn.Add(oData);
    '    }

    '    return listReturn;
    '}

    'public virtual List<T> LoadByQuery(string hsqlQuery)
    '{

    '    List<T> listReturn = new List<T>();
    '    IList oArray = null;

    '    oArray = nhTemplate.Find(hsqlQuery) as IList;
    '    listReturn = new List<T>();
    '    foreach (T oData in oArray)
    '    {
    '        listReturn.Add(oData);
    '    }

    '    return listReturn;
    '}
#End Region

#Region " IDisposable Support "

    Private disposedValue As Boolean = False        ' To detect redundant calls

    ' IDisposable
    Protected Overridable Sub Dispose(ByVal disposing As Boolean)
        If Not Me.disposedValue Then
            If disposing Then
                If oSession.Transaction.IsActive Then
                    oSession.Transaction.Commit()
                End If
                oTransaction.Dispose()
                oSession.Close()
                oSession.Dispose()
            End If
            ' TODO: free shared unmanaged resources
        End If
        Me.disposedValue = True
    End Sub

    ' This code added by Visual Basic to correctly implement the disposable pattern.
    Public Sub Dispose() Implements IDisposable.Dispose
        ' Do not change this code.  Put cleanup code in Dispose(ByVal disposing As Boolean) above.
        Dispose(True)
        GC.SuppressFinalize(Me)
    End Sub


#End Region

End Class

File: Jack.Data/Repository/GenericDao.cs
using NHibernate;
using System;
using System.Collections.Generic;

namespace Jack.Data
{
    public class GenericDao<Tipo, ID> : IGenericDao<Tipo, ID>, IDisposable
    {

        public ISession oSession;
        private ITransaction oTransaction;
        private bool blnHasTransaction;
        //Private oNhTemplate As 

        /// <summary>
        /// Construtor, obtém a sessão do NHibernateHelper e inicia a transação.
        /// </summary>
        public GenericDao(bool blnWithTransaction = false)
        {
            try
            {
                Type tipoEntidade = typeof(Tipo);
                oSession = NHibernateHelper.GetCurrentSession(tipoEntidade.Assembly);
                oSession.GetSessionImplementation().PersistenceContext.Unproxy(tipoEntidade.Assembly);
                blnHasTransaction = blnWithTransaction;
                if (blnWithTransaction)
                {
                    oTransaction = oSession.BeginTransaction();
                }
            }
            catch (Exception ex)
            {
                throw ex;
            }
        }

        /// <summary>
        /// Persiste um novo objeto.
        /// </summary>
        /// <param name="Obj">Objeto a ser persistido</param>
        /// <returns>True para sucesso, False para erro.</returns>
        public virtual bool Insert(Tipo Obj)
        {
            object oRetorno = null;
            try
            {
                oRetorno = oSession.Save(Obj);
                Finish();
            }
            catch (Exception ex)
            {
                throw ex;
            }
            return true;
        }

        /// <summary>
        /// Atualiza um objeto existente.
        /// </summary>
        /// <param name="Obj">Objeto a ser atualizado</param>
        /// <returns>True para sucesso, False para erro.</returns>
        public virtual bool Update(Tipo Obj)
        {
            try
            {
                oSession.Update(Obj);
                Finish();
            }
            catch (Exception ex)
            {
                throw ex;
            }
            return true;
        }

        /// <summary>
        /// Exclui um objeto existente.
        /// </summary>
        /// <param name="Obj">Objeto a ser atualizado</param>
        /// <returns>True para sucesso, False para erro.</returns>
        public virtual bool Delete(Tipo Obj)
        {
            try
            {
                oSession.Delete(Obj);
                Finish();
            }
            catch (Exception ex)
            {
                throw ex;
            }
            return true;
        }

        /// <summary>
        /// Recupera um objeto persistido na base de dados.
        /// </summary>
        /// <param name="Identifier">Chave primária do objeto</param>
        /// <returns>Objeto</returns>
        public virtual Tipo Find(ID Identifier)
        {
            Tipo oTipo = default(Tipo);

            try
            {
                //oTipo = (Tipo)oSession.Load(typeof(Tipo), Identifier);
                oTipo = oSession.Get<Tipo>(Identifier);
                return oTipo;
            }
            catch (Exception ex)
            {
                throw ex;
            }
        }

        /// <summary>
        /// Retorna todos os objetos persistidos na base de dados.
        /// </summary>
        /// <returns>Coleção de objetos</returns>
        public virtual IList<Tipo> LoadAll()
        {
            IList<Tipo> lTipo = default(IList<Tipo>);
            try
            {
                //Prepare()
                lTipo = oSession.CreateCriteria(typeof(Tipo)).List<Tipo>();
                return lTipo;
            }
            catch (Exception ex)
            {
                throw ex;
            }
        }

        /// <summary>
        /// Confirma a transação corrente (commit)
        /// </summary>
        protected virtual void Commit()
        {
            oTransaction.Commit();
        }

        /// <summary>
        /// Desfaz a transação corrente (rollback)
        /// </summary>
        protected virtual void BeginTransaction()
        {
            oTransaction = oSession.BeginTransaction();
        }

        /// <summary>
        /// Begin Trans
        /// </summary>
        /// <remarks></remarks>
        protected virtual void Rollback()
        {
            oTransaction.Rollback();
        }

        /// <summary>
        /// Reinicia a Sessão
        /// </summary>
        /// <remarks></remarks>
        protected virtual void Finish()
        {
            oSession.Flush();
            oSession.Clear();
        }

        /// <summary>
        /// Verifica se há ou não transação
        /// </summary>
        /// <value>Boolean</value>
        /// <returns>Boolean</returns>
        /// <remarks></remarks>
        public bool HasTransaction
        {
            get { return blnHasTransaction; }
            set { blnHasTransaction = value; }
        }

        /// <summary>
        /// Retorna dados através de uma query
        /// </summary>
        /// <param name="strQuery">Query</param>
        /// <param name="strValue">Valor a Filtrar</param>
        /// <param name="strParName">Nome do Parametro</param>
        /// <returns>Lista Tipada</returns>
        /// <remarks></remarks>
        public IList<Tipo> LoadByQuery(string strQuery, string strParName, string strValue)
        {

            IList<Tipo> oReturn = default(IList<Tipo>);
            NHibernate.IQuery oQuery = default(NHibernate.IQuery);

            try
            {
                oQuery = oSession.CreateQuery(strQuery).SetString(strParName, strValue);
                oReturn = oQuery.List<Tipo>();

            }
            catch (Exception ex)
            {
                oReturn = null;
                throw ex;
            }

            return oReturn;

        }


        #region "other Methods"

        //public virtual List<Tipo> LoadByQuery(string hsqlQuery, object[] values)
        //{

        //    List<Tipo> listReturn = new List<Tipo>();
        //    IList oArray = null;

        //    oArray = this.Find(hsqlQuery, values) as IList;
        //    listReturn = new List<Tipo>();
        //    foreach (Tipo oData in oArray)
        //    {
        //        listReturn.Add(oData);
        //    }

        //    return listReturn;
        //}

        //public virtual List<Tipo> LoadByQuery(string hsqlQuery)
        //{

        //    List<Tipo> listReturn = new List<Tipo>();
        //    IList oArray = null;

        //    oArray = this.Find(hsqlQuery) as IList;
        //    listReturn = new List<Tipo>();
        //    foreach (Tipo oData in oArray)
        //    {
        //        listReturn.Add(oData);
        //    }

        //    return listReturn;
        //}
        #endregion

        #region " IDisposable Support "

        // To detect redundant calls
        private bool disposedValue = false;

        // IDisposable
        protected virtual void Dispose(bool disposing)
        {
            if (!this.disposedValue)
            {
                if (disposing)
                {
                    if (oSession.Transaction.IsActive)
                    {
                        oSession.Transaction.Commit();
                    }
                    oTransaction.Dispose();
                    oSession.Close();
                    oSession.Dispose();
                }
                // TODO: free shared unmanaged resources
            }
            this.disposedValue = true;
        }

        // This code added by Visual Basic to correctly implement the disposable pattern.
        public void Dispose()
        {
            // Do not change this code.  Put cleanup code in Dispose(ByVal disposing As Boolean) above.
            Dispose(true);
            GC.SuppressFinalize(this);
        }


        #endregion

    }
}
